### PR TITLE
Add documentation for all Tensor related classes/functions.

### PR DESCRIPTION
**Pull request by xyzsam**

### CODE REVIEW

File: smaug/core/tensor.h [hunk 1]
```diff
@@ -17,6 +17,21 @@
 
 namespace smaug {
 
+/**
```
Comment (yaoyuannnn) [inline smaug/core/tensor.h:20]: Just curious, do you just any VIM shortcut to auto-comment a code block?

File: smaug/core/tensor.h [hunk 2]
```diff
@@ -360,8 +443,12 @@ class Tensor : public TensorBase {
 #endif
     }
 
-    // Fill the tensor with float16 data. This is needed because the data in
-    // tensor proto packs two float16 into one int32.
+    /**
+     * Fill the tensor with float16 data.
+     *
+     * This special overload is required is needed because the data stored in
```
Comment (yaoyuannnn) [inline smaug/core/tensor.h:449]: Remove `is needed`.

File: smaug/core/tensor.h [hunk 3]
```diff
@@ -414,38 +511,59 @@ class Tensor : public TensorBase {
...
     }
 
-    // Return a TensorProto that serializes this Tensor.
+    /** Serializes this Tensor to a TensorProto. */
     TensorProto* asTensorProto();
 
+    /**
+     * Returns a const pointer to the Tensor data.
+     */
+    // TODO(xyzsam): Should this be const T*?
```
Comment (yaoyuannnn) [inline smaug/core/tensor.h:520]: Yeah, I think here we want to return a pointer to const T data. So this should be `const T*`.

File: smaug/core/tensor.h [hunk 4]
```diff
@@ -414,38 +511,59 @@ class Tensor : public TensorBase {
...
            : TensorBase(), origTensor(_origTensor), useRawTensor(_useRawTensor),
              dataFilled(false) {}
+   /**
+    * Construct a TiledTensor.
+    *
+    * @param shape The shape of the TiledTensor - that is, the number of tiles
+    * in each dimension. Alignment padding is ignored here.
+    * @param _origTensor The Tensor that is being tiled.
+    * @param _useRawTensor If true, data from the original Tensor is memcpy'ed
+    * into the all thetiles, instead of being copied element-wise from/to a
```
Comment (yaoyuannnn) [inline smaug/core/tensor.h:564]: `the tiles`.

File: smaug/core/tensor.h (hunk 1 above)
Comment (xyzsam) [reply smaug/core/tensor.h:20]: Nope, all done by hand :(

File: smaug/core/tensor.h (hunk 2 above)
Comment (xyzsam) [reply smaug/core/tensor.h:449]: Done.

File: smaug/core/tensor.h (hunk 3 above)
Comment (xyzsam) [reply smaug/core/tensor.h:520]: I'll put all code-related fixes into a separate commit so that this commit only deals with docs.

**Pull request merged**